File: src/main/java/dev/lbuddyboy/flash/rank/Rank.java
package dev.lbuddyboy.flash.rank;

import dev.lbuddyboy.flash.Flash;
import dev.lbuddyboy.flash.rank.packet.RanksUpdatePacket;
import dev.lbuddyboy.flash.user.User;
import lombok.Data;
import org.bukkit.ChatColor;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Data
public abstract class Rank {

    public UUID uuid;
    public String name;
    public String displayName = name;
    public ChatColor color = ChatColor.WHITE;
    public int weight = 1;
    public String prefix = "";
    public String suffix = "";
    public List<String> permissions = new ArrayList<>();
    public List<String> inheritance = new ArrayList<>();
    public boolean defaultRank, staff;

    public abstract void load();
    public abstract void delete();
    public abstract void save(boolean async);

    public List<User> getUsersWithRank() {
        List<User> peopleWithThisRank = new ArrayList<>();

        for (UUID uuid : Flash.getInstance().getCacheHandler().getUserCache().allUUIDs()) {
            User user = Flash.getInstance().getUserHandler().tryUserRank(uuid, true);

            if (user == null) continue;

            if (user.getActiveRank() != null && user.getActiveRank().getUuid().toString().equalsIgnoreCase(this.uuid.toString())) peopleWithThisRank.add(user);
        }

        return peopleWithThisRank;
    }

    public String getColoredName() {
        return this.color + this.name;
    }

    public String getDisplayName() {
        if (displayName == null) return this.name;

        return this.displayName;
    }

    public List<String> getInheritedPermissions() {
        List<String> perms = new ArrayList<>();

        for (String key : this.inheritance) {
            Rank rank = Flash.getInstance().getRankHandler().getRank(key);
            if (rank == null) continue;
            perms.addAll(rank.getPermissions());
        }

        return perms;
    }

    public void setDefaultRank() {
        for (Rank rank : Flash.getInstance().getRankHandler().getRanks().values()) {
            if (rank.isDefaultRank()) {
                rank.setDefaultRank(false);
            }
        }

        this.defaultRank = !defaultRank;
        Flash.getInstance().getRankHandler().setDefaultRank(this);

        new RanksUpdatePacket(Flash.getInstance().getRankHandler().getRanks()).send();
    }

}

File: src/main/java/dev/lbuddyboy/flash/rank/Rank.kt
package dev.lbuddyboy.flash.rank

import dev.lbuddyboy.flash.Flashimport

dev.lbuddyboy.flash.rank.packet.RanksUpdatePacketimport dev.lbuddyboy.flash.user.Userimport lombok.*import org.bukkit.*
import java.util.*

@Data
abstract class Rank {
    var uuid: UUID? = null
    var name: String? = null
    var displayName = name
    var color = ChatColor.WHITE
    var weight = 1
    var prefix = ""
    var suffix = ""
    var permissions: List<String> = ArrayList()
    var inheritance: List<String> = ArrayList()
    var defaultRank = false
    var staff = false
    abstract fun load()
    abstract fun delete()
    abstract fun save(async: Boolean)
    val usersWithRank: List<User>
        get() {
            val peopleWithThisRank: MutableList<User> = ArrayList()
            for (uuid in Flash.instance.cacheHandler.getUserCache().allUUIDs()) {
                val user = Flash.instance.userHandler.tryUserRank(uuid, true) ?: continue
                if (user.activeRank != null && user.activeRank.getUuid().toString()
                        .equals(this.uuid.toString(), ignoreCase = true)
                ) peopleWithThisRank.add(user)
            }
            return peopleWithThisRank
        }
    val coloredName: String
        get() = color.toString() + name

    fun getDisplayName(): String? {
        return if (displayName == null) name else displayName
    }

    val inheritedPermissions: List<String?>
        get() {
            val perms: MutableList<String?> = ArrayList()
            for (key in inheritance) {
                val rank = Flash.instance.rankHandler.getRank(key) ?: continue
                perms.addAll(rank.getPermissions())
            }
            return perms
        }

    fun setDefaultRank() {
        for (rank in Flash.instance.rankHandler.getRanks().values()) {
            if (rank.isDefaultRank()) {
                rank.setDefaultRank(false)
            }
        }
        defaultRank = !defaultRank
        Flash.instance.rankHandler.setDefaultRank(this)
        RanksUpdatePacket(Flash.instance.rankHandler.getRanks()).send()
    }
}